core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java

```diff
@@ -0,0 +1,216 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import tachyon.Constants;
+
+/**
+ * <p>
+ * Central registry of available {@link UnderFileSystemFactory} instances that uses the
+ * {@link ServiceLoader} mechanism to automatically discover available factories and provides a
+ * central place for obtaining actual {@link UnderFileSystem} instances.
+ * </p>
+ * <h3>Registering New Factories</h3>
+ * <p>
+ * New factories can be registered either using the {@linkplain ServiceLoader} based automatic
+ * discovery mechanism or manually using the static {@link #add(UnderFileSystemFactory)} method. The
+ * down-side of the automatic discovery mechanism is that the discovery order is not controllable so
+ * if your implementation is designed as a replacement for one of the standard implementations
+ * depending on the order in which the JVM discovers the services your own implementation may not
+ * take priority. You can enable {@code DEBUG} level logging for this class to see the order in
+ * which factories are discovered and which is selected when obtaining a {@link UnderFileSystem}
+ * instance for a path. If this shows that your implementation is not getting discovered or used
+ * then you may wish to use the manual registration approach.
+ * <p/>
+ * <h4>Automatic Discovery</h4>
+ * <p>
+ * To use the {@linkplain ServiceLoader} based mechanism you need to have a file named
+ * {@code tachyon.underfs.UnderFileSystemFactory} placed in the {@code META-INF\services} directory
+ * of your project. This file should contain the full name of your factory types (one per line),
+ * your factory types must have a public unparameterised constructor available (see
+ * {@link ServiceLoader} for more detail on this). You can enable {@code DEBUG} level logging to see
+ * factories as they are discovered if you wish to check that your implementation gets discovered.
+ * </p>
+ * <p>
+ * Note that if you are bundling Tachyon plus your code in a shaded JAR using Maven make sure to use
+ * the {@code ServicesResourceTransformer} as otherwise your services file will override the core
+ * provided services file and leave the standard factories and under file system implementations
+ * unavailable.
+ * </p>
+ * <h4>Manual Registration</h4>
+ * <p>
+ * To manually register a factory simply pass an instance of your factory to the
+ * {@link #add(UnderFileSystemFactory)} method. This can be useful when your factory cannot be
+ * instantiated without arguments or in cases where automatic discovery does not give your factory
+ * priority. Factories registered this way will be registered at the start of the factories list so
+ * will have the first opportunity to indicate whether they support a requested path.
+ * </p>
+ */
+public class UnderFileSystemRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final List<UnderFileSystemFactory> FACTORIES =
```

Comment:
Might be better to use a map with key of schema.  That way find is just a check in the map rather than iterate over.